# Introduction to Pandas

## Pandas provides Python data frames

* Popular and established
* Inspired by R dataframes
* Built on `numpy` for fast computation

In [1]:
import pandas as pd

## Our first dataframe

In [2]:
df = pd.DataFrame({"Names": ["Iverson", "Malone", "Bergen"],
                   "Python_mastery": [10, 5, 1.0],
                   "Love_of_R": [2, 5, 11],
                   "years_at_wsu": [4, 17, 5]})
df.head()

Names  Python_mastery  Love_of_R  years_at_wsu
0  Iverson            10.0          2             4
1   Malone             5.0          5            17
2   Bergen             1.0         11             5

## Reading from a csv

* Most data sets will be read in from a csv or JSON data file
* `Pandas` provides `read_csv` and `read_json`

### Open a local file w/ relative path

In [42]:
# Won't work in colab
artist = pd.read_csv('./data/Artists.csv')
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992      None          NaN  
1       1936        0      None          NaN  
2       1941        0      None          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0      None          NaN

### Open a web address

In [43]:
url = "https://github.com/MuseumofModernArt/collection/raw/master/Artists.csv"
artists =  pd.read_csv(url)
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992       NaN          NaN  
1       1936        0       NaN          NaN  
2       1941        0       NaN          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0       NaN          NaN

# JSON data file

* Another (more modern) storage
* Here the data is stored in row `dict`

```{json}
[
{
  "ConstituentID": 1,
  "DisplayName": "Robert Arneson",
  "ArtistBio": "American, 1930–1992",
  "Nationality": "American",
  "Gender": "Male",
  "BeginDate": 1930,
  "EndDate": 1992,
  "Wiki QID": null,
  "ULAN": null
},
{
  "ConstituentID": 2,
  "DisplayName": "Doroteo Arnaiz",
  "ArtistBio": "Spanish, born 1936",
  "Nationality": "Spanish",
  "Gender": "Male",
  "BeginDate": 1936,
  "EndDate": 0,
  "Wiki QID": null,
  "ULAN": null
},
...
```

## `pandas` can read `json` data

In [44]:
# Won't work in colab
artists =  pd.read_json('./data/Artists.json')
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992      None          NaN  
1       1936        0      None          NaN  
2       1941        0      None          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0      None          NaN

In [45]:
json_url = "https://github.com/MuseumofModernArt/collection/raw/master/Artists.json"
artists =  pd.read_json(json_url)
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992      None          NaN  
1       1936        0      None          NaN  
2       1941        0      None          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0      None          NaN

## <font color="red"> Exercise 2.1.2 </font>
    
Use tab-completion and `help` to discover and explore two more methods of reading a file into a `Pandas` dataframe.


In [ ]:
pd.read_excel

In [ ]:
pd.read_html

In [ ]:
pd.read_json

In [ ]:
pd.read_sql_table

When doing tab completion for pd.read_ I found a couple of interesting things we can read in using pandas. This includes excel, html files, json files, and even sql tables!

## <font color="red"> Exercise 2.1.2 </font>
    
Read in the `Artwork.csv` from [https://github.com/MuseumofModernArt/collection](https://github.com/MuseumofModernArt/collection) and display the head of the resulting dataframe.


In [12]:
artwork = pd.read_csv('https://github.com/MuseumofModernArt/collection/raw/master/Artworks.csv')

In [13]:

artwork.head()

Title  \
0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1  City of Music, National Superior Conservatory ...   
2  Villa near Vienna Project, Outside Vienna, Aus...   
3  The Manhattan Transcripts Project, New York, N...   
4  Villa, project, outside Vienna, Austria, Exter...   

                     Artist ConstituentID  \
0               Otto Wagner          6210   
1  Christian de Portzamparc          7470   
2                Emil Hoppe          7605   
3           Bernard Tschumi          7056   
4                Emil Hoppe          7605   

                                   ArtistBio Nationality BeginDate EndDate  \
0                      (Austrian, 1841–1918)  (Austrian)    (1841)  (1918)   
1                        (French, born 1944)    (French)    (1944)     (0)   
2                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   
3  (French and Swiss, born Switzerland 1944)          ()    (1944)     (0)   
4                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   

   Gender  Date                                             Medium  ...  \
0  (Male)  1896      Ink and cut-and-pasted painted pages on paper  ...   
1  (Male)  1987                  Paint and colored pencil on print  ...   
2  (Male)  1903  Graphite, pen, color pencil, ink, and gouache ...  ...   
3  (Male)  1980  Photographic reproduction with colored synthet...  ...   
4  (Male)  1903  Graphite, color pencil, ink, and gouache on tr...  ...   

                                        ThumbnailURL Circumference (cm)  \
0  http://www.moma.org/media/W1siZiIsIjU5NDA1Il0s...                NaN   
1  http://www.moma.org/media/W1siZiIsIjk3Il0sWyJw...                NaN   
2  http://www.moma.org/media/W1siZiIsIjk4Il0sWyJw...                NaN   
3  http://www.moma.org/media/W1siZiIsIjEyNCJdLFsi...                NaN   
4  http://www.moma.org/media/W1siZiIsIjEyNiJdLFsi...                NaN   

  Depth (cm) Diameter (cm) Height (cm) Length (cm) Weight (kg)  Width (cm)  \
0        NaN           NaN     48.6000         NaN         NaN    168.9000   
1        NaN           NaN     40.6401         NaN         NaN     29.8451   
2        NaN           NaN     34.3000         NaN         NaN     31.8000   
3        NaN           NaN     50.8000         NaN         NaN     50.8000   
4        NaN           NaN     38.4000         NaN         NaN     19.1000   

  Seat Height (cm) Duration (sec.)  
0              NaN             NaN  
1              NaN             NaN  
2              NaN             NaN  
3              NaN             NaN  
4              NaN             NaN  

[5 rows x 29 columns]

## So what is a `DateFrame`

* Like R, Pandas focuses on columns
* Think `dict` of `(str, Series)` pairs 
* A series is a typed list-like structure

In [4]:
# This is how I imagine a dataframe
df = pd.DataFrame({"Names": ["Iverson", "Malone", "Bergen"],
                   "Python_mastery": [10, 5, 1.0],
                   "years_at_wsu": [4.5, 17.5, 5.5]})

In [5]:
type(df)

pandas.core.frame.DataFrame

## Columns are `Series` and hold one type of data

In [63]:
type(artists.BeginDate), type(artists.DisplayName)

(pandas.core.series.Series, pandas.core.series.Series)

In [64]:
artists.BeginDate.dtype, artists.DisplayName.dtype

(dtype('int64'), dtype('O'))

## Two ways to access a column

* **Method 1:** like a dictionary
    * `df["column_name"]`
* **Method 2:** like an object attribute
    * `df.column_name`
    * Only for proper names!

In [65]:
artists.BeginDate.head(2)

0    1930
1    1936
Name: BeginDate, dtype: int64

In [66]:
artists['BeginDate'].head(2)

0    1930
1    1936
Name: BeginDate, dtype: int64

## More on data types

* See all data types with `df.dtypes`
* You can set the `dtypes` when you read a dataframe
* Read more about types: [Pandas docs](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.dtypes.html)

In [52]:
artists.dtypes

ConstituentID      int64
DisplayName       object
ArtistBio         object
Nationality       object
Gender            object
BeginDate          int64
EndDate            int64
Wiki QID          object
ULAN             float64
dtype: object

## Setting `dtypes` with `read_csv`

We can pass a `dict` of types to `dtype` keyword

In [56]:
import numpy as np
artist_types = {'ConstituentID': np.int64,
                'DisplayName': str,
                'ArtistBio': str,
                'Nationality': str,
                'Gender':str,
                'BeginDate': np.int64,
                'EndDate': np.int64,
                'Wiki QID': str,
                'ULAN':pd.Int64Dtype()} # If you get an error ==> update pandas (see below)
artists2 = pd.read_csv('./data/artists.csv', dtype = artist_types)
artists2.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID       ULAN  
0       1930     1992       NaN       <NA>  
1       1936        0       NaN       <NA>  
2       1941        0       NaN       <NA>  
3       1946        0  Q1063584  500027998  
4       1941        0       NaN       <NA>

## What's up with `ULAN` ?

* Currently, `numpy` $\rightarrow$ no missing `int`s
* Pandas correct this with the `pd.Int64Dtype()` type
    * Only available in `pandas >= 0.24.0`

In [57]:
pd.__version__

'1.1.3'

## An `Int` by any other name ...

* `np.int64` $\rightarrow$ no missing values
* `pd.Int64Dtype()` $\rightarrow$ allows `NaN`

In [58]:
artists2.dtypes

ConstituentID     int64
DisplayName      object
ArtistBio        object
Nationality      object
Gender           object
BeginDate         int64
EndDate           int64
Wiki QID         object
ULAN              Int64
dtype: object

## Preview of coming attractions

* Now we can switch `BeginDate` and `EndDate` from `0` to `np.NaN`
* We will do this in the next section

# Getting to know your data

## Basic inspection tools

* `df.head()`        first five rows
* `df.tail()`        last five rows
* `df.sample(5)`     random sample of rows
* `df.shape`         number of rows/columns in a tuple
* `df.describe()`    calculates measures of central tendency
* `df.info()`

## <font color="red"> Exercise 1: Inspect the artwork from MoMA </font>

#### Read the csv and inspect the `head`

In [14]:
artwork.head()

Title  \
0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1  City of Music, National Superior Conservatory ...   
2  Villa near Vienna Project, Outside Vienna, Aus...   
3  The Manhattan Transcripts Project, New York, N...   
4  Villa, project, outside Vienna, Austria, Exter...   

                     Artist ConstituentID  \
0               Otto Wagner          6210   
1  Christian de Portzamparc          7470   
2                Emil Hoppe          7605   
3           Bernard Tschumi          7056   
4                Emil Hoppe          7605   

                                   ArtistBio Nationality BeginDate EndDate  \
0                      (Austrian, 1841–1918)  (Austrian)    (1841)  (1918)   
1                        (French, born 1944)    (French)    (1944)     (0)   
2                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   
3  (French and Swiss, born Switzerland 1944)          ()    (1944)     (0)   
4                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   

   Gender  Date                                             Medium  ...  \
0  (Male)  1896      Ink and cut-and-pasted painted pages on paper  ...   
1  (Male)  1987                  Paint and colored pencil on print  ...   
2  (Male)  1903  Graphite, pen, color pencil, ink, and gouache ...  ...   
3  (Male)  1980  Photographic reproduction with colored synthet...  ...   
4  (Male)  1903  Graphite, color pencil, ink, and gouache on tr...  ...   

                                        ThumbnailURL Circumference (cm)  \
0  http://www.moma.org/media/W1siZiIsIjU5NDA1Il0s...                NaN   
1  http://www.moma.org/media/W1siZiIsIjk3Il0sWyJw...                NaN   
2  http://www.moma.org/media/W1siZiIsIjk4Il0sWyJw...                NaN   
3  http://www.moma.org/media/W1siZiIsIjEyNCJdLFsi...                NaN   
4  http://www.moma.org/media/W1siZiIsIjEyNiJdLFsi...                NaN   

  Depth (cm) Diameter (cm) Height (cm) Length (cm) Weight (kg)  Width (cm)  \
0        NaN           NaN     48.6000         NaN         NaN    168.9000   
1        NaN           NaN     40.6401         NaN         NaN     29.8451   
2        NaN           NaN     34.3000         NaN         NaN     31.8000   
3        NaN           NaN     50.8000         NaN         NaN     50.8000   
4        NaN           NaN     38.4000         NaN         NaN     19.1000   

  Seat Height (cm) Duration (sec.)  
0              NaN             NaN  
1              NaN             NaN  
2              NaN             NaN  
3              NaN             NaN  
4              NaN             NaN  

[5 rows x 29 columns]

**Task:** Write a few sentences describing an problems

Some issues in this data set is that data is surrounded in parenthesis, have commas, are very long so not spread apart (bad data managment).

#### Inspect the column names with the `columns` attribute

In [15]:
artwork.columns

Index(['Title', 'Artist', 'ConstituentID', 'ArtistBio', 'Nationality',
       'BeginDate', 'EndDate', 'Gender', 'Date', 'Medium', 'Dimensions',
       'CreditLine', 'AccessionNumber', 'Classification', 'Department',
       'DateAcquired', 'Cataloged', 'ObjectID', 'URL', 'ThumbnailURL',
       'Circumference (cm)', 'Depth (cm)', 'Diameter (cm)', 'Height (cm)',
       'Length (cm)', 'Weight (kg)', 'Width (cm)', 'Seat Height (cm)',
       'Duration (sec.)'],
      dtype='object')

**Question:** See any problems?

Most of the columns look to be in good shape but the columns that have (cm) in their name may cause issues down the road. But this could be worked around by using the df['column name'] format

#### Inspect the tail

In [16]:
artwork.tail()

Title  \
138146                                           Untitled   
138147  Plate (folio 2 verso) from Muscheln und schirm...   
138148  Plate (folio 6) from Muscheln und schirme (She...   
138149  Plate (folio 12) from Muscheln und schirme (Sh...   
138150  Plate (folio 16) from Muscheln und schirme (Sh...   

                                                   Artist  \
138146  Chesnutt Brothers Studio, Andrew Chesnutt, Lew...   
138147                                 Sophie Taeuber-Arp   
138148                                 Sophie Taeuber-Arp   
138149                                 Sophie Taeuber-Arp   
138150                                 Sophie Taeuber-Arp   

                 ConstituentID                                    ArtistBio  \
138146  133005, 133006, 133007  (American, 1861–1934) (American, 1860–1933)   
138147                    5777                           (Swiss, 1889–1943)   
138148                    5777                           (Swiss, 1889–1943)   
138149                    5777                           (Swiss, 1889–1943)   
138150                    5777                           (Swiss, 1889–1943)   

                     Nationality          BeginDate            EndDate  \
138146  () (American) (American)  (0) (1861) (1860)  (0) (1934) (1933)   
138147                   (Swiss)             (1889)             (1943)   
138148                   (Swiss)             (1889)             (1943)   
138149                   (Swiss)             (1889)             (1943)   
138150                   (Swiss)             (1889)             (1943)   

                  Gender     Date  \
138146  () (Male) (Male)  c. 1890   
138147          (Female)     1939   
138148          (Female)     1939   
138149          (Female)     1939   
138150          (Female)     1939   

                                                   Medium  ...  \
138146                               Gelatin silver print  ...   
138147  One from an illustrated book with four line bl...  ...   
138148  One from an illustrated book with four line bl...  ...   
138149  One from an illustrated book with four line bl...  ...   
138150  One from an illustrated book with four line bl...  ...   

                                             ThumbnailURL Circumference (cm)  \
138146  http://www.moma.org/media/W1siZiIsIjQ5MjcyMiJd...                NaN   
138147  http://www.moma.org/media/W1siZiIsIjQ4NTExNSJd...                NaN   
138148  http://www.moma.org/media/W1siZiIsIjQ4NTExOCJd...                NaN   
138149  http://www.moma.org/media/W1siZiIsIjQ4NTEyMCJd...                NaN   
138150  http://www.moma.org/media/W1siZiIsIjQ4NTEyMSJd...                NaN   

       Depth (cm) Diameter (cm) Height (cm) Length (cm) Weight (kg)  \
138146        NaN           NaN   10.795022         NaN         NaN   
138147        NaN           NaN   13.500000         NaN         NaN   
138148        NaN           NaN   13.500000         NaN         NaN   
138149        NaN           NaN   11.000000         NaN         NaN   
138150        NaN           NaN   13.500000         NaN         NaN   

        Width (cm) Seat Height (cm) Duration (sec.)  
138146   16.510033              NaN             NaN  
138147   10.000000              NaN             NaN  
138148   10.000000              NaN             NaN  
138149   10.000000              NaN             NaN  
138150   10.000000              NaN             NaN  

[5 rows x 29 columns]

#### Check out the `shape`

In [17]:
artwork.shape

(138151, 29)

**Question:** What do these number mean?

This shows the demensions of the data or data frame you read in. So for this artwork data it has 138151 rows and 29 columns.

#### Use `describe` to compute statistics

In [18]:
artwork.describe()

ObjectID  Circumference (cm)    Depth (cm)  Diameter (cm)  \
count  138151.000000           10.000000  13839.000000    1462.000000   
mean    97170.256618           44.868020     16.353863      23.094845   
std     81950.720570           28.631604     54.495960      44.626483   
min         2.000000            9.900000      0.000000       0.635000   
25%     36671.500000           23.500000      0.000000       7.778800   
50%     73896.000000           36.000000      0.317501      13.652500   
75%    141636.500000           71.125000      9.842510      24.980950   
max    419289.000000           83.800000   1808.483617     914.400000   

         Height (cm)  Length (cm)    Weight (kg)     Width (cm)  \
count  120355.000000   742.000000     290.000000  119434.000000   
mean       37.456124    89.687579    1283.674965      37.973398   
std        49.604159   329.428165   12017.504240      67.277097   
min         0.000000     0.000000       0.090000       0.000000   
25%        17.938786    17.100000       5.726700      17.500000   
50%        27.800000    26.700000      20.185100      25.400051   
75%        43.900000    79.700000      77.678500      44.200000   
max      9140.000000  8321.056600  185067.585957    9144.000000   

       Seat Height (cm)  Duration (sec.)  
count               0.0     2.140000e+03  
mean                NaN     6.156488e+03  
std                 NaN     1.371250e+05  
min                 NaN     0.000000e+00  
25%                 NaN     1.200000e+02  
50%                 NaN     4.330000e+02  
75%                 NaN     1.620000e+03  
max                 NaN     6.283065e+06

#### Use `info` to look at types and totals

In [19]:
artwork.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138151 entries, 0 to 138150
Data columns (total 29 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Title               138112 non-null  object 
 1   Artist              136868 non-null  object 
 2   ConstituentID       136868 non-null  object 
 3   ArtistBio           132226 non-null  object 
 4   Nationality         136868 non-null  object 
 5   BeginDate           136868 non-null  object 
 6   EndDate             136868 non-null  object 
 7   Gender              136868 non-null  object 
 8   Date                135949 non-null  object 
 9   Medium              128450 non-null  object 
 10  Dimensions          128397 non-null  object 
 11  CreditLine          135714 non-null  object 
 12  AccessionNumber     138151 non-null  object 
 13  Classification      138151 non-null  object 
 14  Department          138151 non-null  object 
 15  DateAcquired        131026 non-nul

**Question:** What did you learn from the last two cells?

The last two cells show the statistics of the data like the mean, max, min, counts, etc. The info shows what the columns are or how the columns are structured.  For example the column Title is object Dtype and ObjectID is a int64 Dtype. The last cell also shows how many calues are non-null and the count.